graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -0,0 +1,173 @@
+
+    name_to_schema_id = {}  # Dict[str, str], name of type/interface/enum/union to schema id
+    scalars = {'String', 'Int', 'Float', 'Boolean', 'ID'}  # Set[str], user defined + builtins
+    directives = {}  # Dict[str, DirectiveDefinition]
+
+    for cur_schema_id, cur_ast in six.iteritems(schemas_dict):
+        # Prevent aliasing between output and input
+        cur_ast = deepcopy(cur_ast)
+
+        # Check input schema satisfies various structural requirements
+        check_ast_schema_is_valid(cur_ast)
+
+        cur_schema = build_ast_schema(cur_ast)
+        cur_query_type = get_query_type_name(cur_schema)
+
+        # Merge cur_ast into merged_schema_ast
+        # Concatenate new scalars, new directives, and type definitions other than the query
+        # type to definitions list
+        # Raise errors for conflicting scalars, directives, or types
+        new_definitions = cur_ast.definitions  # List[Node]
+        new_query_type_fields = None  # List[FieldDefinition]
+        for new_definition in new_definitions:
+            if isinstance(new_definition, ast_types.SchemaDefinition):
+                continue
+
+            new_name = new_definition.name.value
+
+            if (
+                isinstance(new_definition, ast_types.ObjectTypeDefinition) and
+                new_name == cur_query_type
+            ):  # query type definition
+                new_query_type_fields = new_definition.fields  # List[FieldDefinition]
+
+            elif isinstance(new_definition, ast_types.DirectiveDefinition):
+                if new_name in directives:  # existing directive
+                    if new_definition == directives[new_name]:  # definitions agree
+                        continue
+                    else:  # definitions disagree
+                        raise SchemaNameConflictError(
+                            u'Directive "{}" with definition "{}" has already been defined with '
+                            u'definition "{}".'.format(new_name, print_ast(new_definition),
+                                                       print_ast(directives[new_name]))
+                        )
+                # new directive
+                merged_schema_ast.definitions.append(new_definition)  # Add to AST
+                directives[new_name] = new_definition
+
+            elif isinstance(new_definition, ast_types.ScalarTypeDefinition):
+                if new_name in scalars:  # existing scalar
+                    continue
+                if new_name in name_to_schema_id:  # new scalar clashing with existing type
+                    raise SchemaNameConflictError(
+                        u'New scalar "{}" clashes with existing type.'.format(new_name)
+                    )
+                # new, valid scalar
+                merged_schema_ast.definitions.append(new_definition)  # Add to AST
+                scalars.add(new_name)
+
+            else:  # Generic type definition
```

Comment:
Can we turn this into an `elif`, so we can positively check that we are indeed at a generic type definition, and turn the `else` case into an assertion error about unreachable code?